aten/src/ATen/native/ComparisonUtils.cpp

```diff
@@ -14,6 +15,20 @@ class Tensor;
 
 namespace native {
 
+template<typename O, typename C>
+void _assert_match_sym(const O& original, const C& compared, const std::string& name) {
+  if (compared) {
+    auto sym_compared = c10::SymIntArrayRef(
+      reinterpret_cast<const c10::SymInt*>(compared.value().data()), compared.value().size());
```

Comment:
This seems way too dangerous. You're doing an unrestricted reinterpret_cast, where you're allowing C to be literally any type.

Reply:
And looking at the call site you are doing the reinterpret cast on an IntArrayRef which is definitely not kosher. There are functions specifically to do this conversion (the problem is an IntArrayRef can have large negative numbers which will pun incorrectly).

Reply:
Also move this shit into its own PR, this can totally stand on its own.